File: bowling-game/python/bowling_game/game.py
class Game(object):
    def __init__(self):
        self._rolls = []
        for i in range(0,21): self._rolls.append(0)
        self._current_roll = 0

    def roll(self, pins):
        self._rolls[self._current_roll] = pins
        self._current_roll +=1

    @property
    def score(self):
        _score = 0
        frame_index =0
        for frame in range(0, 10):
            if self._is_strike(frame_index):
                _score += 10 + self._rolls[frame_index+1] + self._rolls[frame_index+2]
                frame_index += 1
            elif self._is_spare(frame_index):
                _score += 10 + self._rolls[frame_index+2]
                frame_index += 2
            else:
                _score += self._rolls[frame_index] + self._rolls[frame_index+1]
                frame_index += 2
        return _score

    def _is_spare(self, frame_index):
        return self._rolls[frame_index] + self._rolls[frame_index+1] == 10
    def _is_strike(self, frame_index):
        return self._rolls[frame_index] == 10

File: bowling-game/elixir/bowling_game/lib/game.ex
defmodule BowlingGame.Game do
  import BowlingGame.Game.Guards

  @type pin_throw :: integer
  @type game_state :: list(pin_throw)
  @type game_score :: integer

  @spec init() :: game_state
  def init do
    []
  end

  @spec roll(game_state, pin_throw) :: game_state
  def roll(rolled_pins, pins) do
    rolled_pins ++ [pins]
  end

  @spec score(game_state) :: game_score
  def score(rolled_pins) do
    frame_scores(rolled_pins)
    |> Enum.sum()
  end

  defp frame_scores([]), do: []

  defp frame_scores(frames) do
    no_of_throws = number_of_throws(frames)

    first_frame = Enum.take(frames, no_of_throws)
    following_frames = Enum.drop(frames, no_of_throws)

    [single_frame_score(first_frame, following_frames)] ++ frame_scores(following_frames)
  end

  defp number_of_throws([first | _]) when first == 10, do: 1
  defp number_of_throws(_), do: 2

  defp single_frame_score(frame, following_frames) do
    frame_score = Enum.sum(frame)

    frame_score + strike_bonus_if_any(frame, following_frames) +
      spare_bonus_if_any(frame, following_frames)
  end

  defp spare_bonus_if_any(frame, following_frames) do
    is_spare?(frame)
    |> spare_bonus(following_frames)
  end

  defp is_spare?(frame) when length(frame) == 1, do: false
  defp is_spare?(frame), do: Enum.sum(frame) == 10

  defp spare_bonus(_is_spare = false, _), do: 0
  defp spare_bonus(_is_spare = true, [first | _rest]), do: first

  defp strike_bonus_if_any(frame, following_frames) do
    is_strike?(frame)
    |> strike_bonus(following_frames)
  end

  defp is_strike?(frame) when has_multiple_throws?(frame), do: false
  defp is_strike?(frame), do: Enum.sum(frame) == 10

  defp strike_bonus(_is_strike = false, _), do: 0
  defp strike_bonus(_is_strike = true, frames) when is_last_frame?(frames), do: 0
  defp strike_bonus(_is_strike = true, [first | [second | _rest]]), do: first + second
end
